### PR TITLE
Riding component and buckle refactor

### PR DESCRIPTION
**Pull request by Rohesie**
## Changelog
:cl:
add: Humans are now able to fireman-carry their fallen comrades by grabbing them with both hands and click-dragging the sprite as if they were stripping them.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 1]
```diff
@@ -435,7 +435,7 @@
 	destination.mob_size_counter += mob_size
 	stop_pulling()
 	smokecloak_off()
-	destination.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /obj/structure/closet/.proc/resisted_against)
+	destination.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /obj/structure/closet/resisted_against)
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/closets.dm]:
```suggestion
	destination.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, .proc/resisted_against)
```

File: code/game/objects/structures/razorwire.dm [hunk 2]
```diff
@@ -62,10 +62,11 @@
 	entangled_list += M //Add the entangled person to the trapped list.
 	M.entangled_by = src
 	ENABLE_BITFIELD(M.restrained_flags, RESTRAINED_RAZORWIRE)
-	RegisterSignal(M, COMSIG_LIVING_DO_RESIST, .proc/resisted_against)
+	RegisterSignal(M, COMSIG_LIVING_DO_RESIST, /obj/structure/razorwire/resisted_against)
```
Comment (spookydonut) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
	RegisterSignal(M, COMSIG_LIVING_DO_RESIST, .proc/resisted_against)
```

File: code/_onclick/telekinesis.dm [hunk 3]
```diff
@@ -245,7 +245,7 @@ Redefine as needed.
 	last_life_tick = victim.life_tick
 
 	ENABLE_BITFIELD(victim.restrained_flags, RESTRAINED_PSYCHICGRAB)
-	RegisterSignal(victim, list(COMSIG_LIVING_DO_RESIST, COMSIG_LIVING_DO_MOVE_RESIST), .proc/resisted_against)
+	RegisterSignal(victim, list(COMSIG_LIVING_DO_RESIST, COMSIG_LIVING_DO_MOVE_RESIST), /obj/item/tk_grab/shrike/resisted_against)
```
Comment (spookydonut) [inline code/_onclick/telekinesis.dm]:
```suggestion
	RegisterSignal(victim, list(COMSIG_LIVING_DO_RESIST, COMSIG_LIVING_DO_MOVE_RESIST), .proc/resisted_against)
```

File: code/game/objects/structures/crates_lockers/closets.dm (hunk 1 above)
Comment (Rohesie) [reply code/game/objects/structures/crates_lockers/closets.dm]: This suggestion fails to compile.

File: code/modules/mob/living/carbon/inventory.dm [hunk 4]
```diff
@@ -4,7 +4,7 @@
 		drop_all_held_items()
 		stop_pulling()
 		handcuffed = restraints
-		handcuffed.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /obj/item/restraints/.proc/resisted_against)
+		handcuffed.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /obj/item/restraints/resisted_against)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/inventory.dm]:
```suggestion
		handcuffed.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /atom/movable.proc/resisted_against)
```

File: code/modules/mob/living/carbon/inventory.dm [hunk 5]
```diff
@@ -17,7 +17,7 @@
 			if(hud_used?.move_intent)
 				hud_used.move_intent.icon_state = "walking"
 		legcuffed = restraints
-		legcuffed.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /obj/item/restraints/.proc/resisted_against)
+		legcuffed.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /obj/item/restraints/resisted_against)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/inventory.dm]:
```suggestion
		legcuffed.RegisterSignal(src, COMSIG_LIVING_DO_RESIST, /atom/movable.proc/resisted_against)
```

File: code/modules/mob/mob_movement.dm [hunk 6]
```diff
@@ -92,7 +92,7 @@
 		return AIMove(n, direct, L)
 
 	//Check if you are being grabbed and if so attemps to break it
-	if(CHECK_BITFIELD(SEND_SIGNAL(L, COMSIG_LIVING_DO_MOVE_RESIST, L), COMSIG_LIVING_RESIST_SUCCESSFUL))
+	if(CHECK_BITFIELD(SEND_SIGNAL(L, COMSIG_LIVING_DO_MOVE_RESIST), COMSIG_LIVING_RESIST_SUCCESSFUL))
```
Comment (spookydonut) [inline code/modules/mob/mob_movement.dm]:
```suggestion
	if(SEND_SIGNAL(L, COMSIG_LIVING_DO_MOVE_RESIST) & COMSIG_LIVING_RESIST_SUCCESSFUL)
```

File: code/modules/atmospherics/machinery/pipes/heat_exchange/he_pipes.dm [hunk 7]
```diff
@@ -3,7 +3,7 @@
 	var/minimum_temperature_difference = 20
 	var/thermal_conductivity = WINDOW_HEAT_TRANSFER_COEFFICIENT
 	color = "#404040"
-	buckle_lying = 1
+	buckle_lying = -1
```
[the remark this answers is not in the document]
Comment (Rohesie) [reply code/modules/atmospherics/machinery/pipes/heat_exchange/he_pipes.dm]: Same as /tg/

[the remark this answers is not in the document]
Comment (spookydonut) [reply code/modules/atmospherics/machinery/pipes/heat_exchange/he_pipes.dm]: Okay but why is it negative, what does that do?

[the remark this answers is not in the document]
Comment (Rohesie) [reply code/modules/atmospherics/machinery/pipes/heat_exchange/he_pipes.dm]: `var/buckle_lying = -1 //bed-like behaviour, forces mob.lying = buckle_lying if != -1`

[the remark this answers is not in the document]
Comment (spookydonut) [reply code/modules/atmospherics/machinery/pipes/heat_exchange/he_pipes.dm]: Change it to a define

File: code/datums/components/riding.dm [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+		handle_vehicle_layer()
+		handle_vehicle_offsets()
+	else
+		to_chat(user, "<span class='warning'>You'll need the keys in one of your hands to [drive_verb] [AM].</span>")
+
+/datum/component/riding/proc/Unbuckle(atom/movable/buckled_thing)
+	addtimer(CALLBACK(parent, /atom/movable/.proc/unbuckle_mob, buckled_thing), 0, TIMER_UNIQUE)
+
+///////Yes, I said humans. No, this won't end well...//////////
+/datum/component/riding/human
```
Comment (spookydonut) [inline code/datums/components/riding.dm]: why does this need a subtype

File: code/datums/components/riding.dm [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+
+	var/del_on_unbuckle_all = FALSE
+
+/datum/component/riding/Initialize()
+	if(!ismovableatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(parent, COMSIG_MOVABLE_BUCKLE, .proc/vehicle_mob_buckle)
+	RegisterSignal(parent, COMSIG_MOVABLE_UNBUCKLE, .proc/vehicle_mob_unbuckle)
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/vehicle_moved)
+
```
Comment (spookydonut) [inline code/datums/components/riding.dm]: wheres the Destroy()

File: code/datums/components/riding.dm [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+/datum/component/riding/human/force_dismount(mob/living/user, silent)
+	var/atom/movable/AM = parent
+	AM.unbuckle_mob(user)
+	user.knock_down(2)
+	if(!silent)
+		user.visible_message("<span class='warning'>[AM] pushes [user] off of [AM.p_them()]!</span>",
+			"<span class='warning'>[AM] pushes you off of [AM.p_them()]!</span>")
+
+
+/datum/component/riding/proc/equip_buckle_inhands(mob/living/carbon/human/user, amount_required = 1, riding_target_override)
```
Comment (spookydonut) [inline code/datums/components/riding.dm]: why is this below the subtype def

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 11]
```diff
@@ -12,8 +12,8 @@
 	desc = "A mattress seated on a rectangular metallic frame. This is used to support a lying person in a comfortable manner, notably for regular sleep. Ancient technology, but still useful."
 	icon_state = "bed"
 	icon = 'icons/obj/objects.dmi'
-	can_buckle = TRUE
-	buckle_lying = TRUE
+	buckle_flags = CAN_BUCKLE|BUCKLE_PREVENTS_PULL
+	buckle_lying = 90
```
Comment (spookydonut) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm]: i assume this is an angle?

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 12]
```diff
@@ -6,25 +6,36 @@
 	name = "chair"
 	desc = "A rectangular metallic frame sitting on four legs with a back panel. Designed to fit the sitting position, more or less comfortably."
 	icon_state = "chair"
-	buckle_lying = FALSE
+	buckle_lying = 0
```
Comment (spookydonut) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: define?

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 13]
```diff
@@ -246,7 +261,7 @@
 /obj/structure/bed/chair/dropship/passenger/rotate()
 	return // no
 
-/obj/structure/bed/chair/dropship/passenger/buckle_mob(mob/living/M, mob/living/user)
+/obj/structure/bed/chair/dropship/passenger/buckle_mob(mob/living/buckling_mob, force = FALSE, check_loc = TRUE, lying_buckle = FALSE, hands_needed = 0, target_hands_needed = 0, silent)
```
Comment (spookydonut) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: thats a lot of arguments

File: code/modules/mob/living/carbon/human/human.dm [hunk 14]
```diff
@@ -787,6 +787,69 @@
...
+	if(buckled || (buckling_mob in buckled_mobs) || (LAZYLEN(buckled_mobs) >= max_buckled_mobs))
+		return
+	var/equipped_hands_self
+	var/equipped_hands_target
+	if(hands_needed)
+		equipped_hands_self = riding_datum.equip_buckle_inhands(src, hands_needed, buckling_mob)
+	if(target_hands_needed)
+		equipped_hands_target = riding_datum.equip_buckle_inhands(buckling_mob, target_hands_needed)
+
+	if(hands_needed || target_hands_needed)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/human/human.dm]: get rid of this if the nested if/else if chain does all the work here

File: code/modules/mob/living/living.dm [hunk 15]
```diff
@@ -249,15 +264,17 @@
...
-	visible_message("<span class='danger'>[src] struggles to break free of [pulledby]'s grip!</span>", null, null, 5)
+	if(pulledby.grab_state >= GRAB_AGGRESSIVE)
+		visible_message("<span class='danger'>[src] struggles to break free of [pulledby]'s grip!</span>", null, null, 5)
 	return resist_grab()
 
 
 /mob/living/resist_grab()
-	if(pulledby.grab_level)
-		if(++grab_resist_level >= pulledby.grab_level)
+	if(pulledby.grab_state)
```
Comment (spookydonut) [inline code/modules/mob/living/living.dm]: early return this

File: code/modules/mob/living/living.dm [hunk 16]
```diff
@@ -249,15 +264,17 @@
...
+		if(++grab_resist_level >= pulledby.grab_state)
```
Comment (spookydonut) [inline code/modules/mob/living/living.dm]: then early return this

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm (hunk 11 above)
Comment (Rohesie) [reply code/game/objects/structures/stool_bed_chair_nest/bed.dm]: Yes, this is how /tg/ does it.

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm (hunk 12 above)
Comment (Rohesie) [reply code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: It's an angle.

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm (hunk 13 above)
Comment (Rohesie) [reply code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: Taken from /tg/

File: code/modules/mob/mob.dm [hunk 17]
```diff
@@ -474,6 +483,39 @@
...
+	if(M.layer < layer)
+		M.layer = layer + 0.1
+///Call back post unbuckle from a mob, (reset your visual height here)
+/mob/post_unbuckle_mob(mob/living/M)
+	M.layer = initial(M.layer)
+	M.pixel_y = initial(M.pixel_y)
+
+///returns the height in pixel the mob should have when buckled to another mob.
+/mob/proc/get_mob_buckling_height(mob/seat)
+	if(isliving(seat))
```
Comment (spookydonut) [inline code/modules/mob/mob.dm]: early return

File: code/datums/components/riding.dm (hunk 8 above)
Comment (Rohesie) [reply code/datums/components/riding.dm]: It's a special case, directly taken from /tg/
It has several redefinitions from the base. Not making it a subtype would require quite the refactor.

File: code/datums/components/riding.dm (hunk 9 above)
Comment (Rohesie) [reply code/datums/components/riding.dm]: There are no variables to clean, nor nothing to be done on Destroy() here.

File: code/modules/vehicles/powerloader.dm [hunk 18]
```diff
@@ -156,14 +161,18 @@
 /obj/item/powerloader_clamp/dropped(mob/user)
 	if(linked_powerloader)
```
Comment (spookydonut) [inline code/modules/vehicles/powerloader.dm]: early return

File: code/datums/components/riding.dm (hunk 9 above)
Comment (spookydonut) [reply code/datums/components/riding.dm]: You can probably make it use Registerwithparent and unregisterwithparent such that you can transfer the component when a xeno evolves

File: code/datums/components/riding.dm [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+/datum/component/riding/proc/Unbuckle(atom/movable/buckled_thing)
+	addtimer(CALLBACK(parent, /atom/movable/.proc/unbuckle_mob, buckled_thing), 0, TIMER_UNIQUE)
+
+///////Yes, I said humans. No, this won't end well...//////////
+/datum/component/riding/human
+	del_on_unbuckle_all = TRUE
+
+/datum/component/riding/human/Initialize()
+	. = ..()
+	RegisterSignal(parent, COMSIG_HUMAN_MELEE_UNARMED_ATTACK, .proc/on_carrier_unarmed_melee)
```
Comment (spookydonut) [inline code/datums/components/riding.dm]: you can just register this in the base type

File: code/datums/components/riding.dm [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+
+/datum/component/riding/human/vehicle_mob_buckle(datum/source, mob/living/buckling_mob, force = FALSE)
+	. = ..()
+	var/mob/living/carbon/human/human_carrier = parent
+	human_carrier.add_movespeed_modifier(MOVESPEED_ID_HUMAN_CARRYING, TRUE, 0, NONE, TRUE, HUMAN_CARRY_SLOWDOWN)
+	RegisterSignal(buckling_mob, COMSIG_MOVABLE_PRE_THROW, .proc/on_passenger_throw)
+
+
+/datum/component/riding/human/proc/on_carrier_unarmed_melee(datum/source, atom/target)
+	var/mob/living/carbon/human/human_carrier = parent
```
Comment (spookydonut) [inline code/datums/components/riding.dm]: this doesnt particularily need to be cast to human, could be on the base type

File: code/datums/components/riding.dm [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+		return
+	force_dismount(target)
+
+/datum/component/riding/human/proc/on_passenger_throw(datum/source, atom/target, range, speed, thrower, spin)
+	force_dismount(source, TRUE)
+
+
+/datum/component/riding/human/handle_vehicle_layer()
+	var/atom/movable/AM = parent
+	if(AM.buckled_mobs && AM.buckled_mobs.len)
```
Comment (spookydonut) [inline code/datums/components/riding.dm]: early return, also this could just use lazylen

File: code/datums/components/riding.dm [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+				AM.layer = OBJ_LAYER
+			else
+				AM.layer = ABOVE_MOB_LAYER
+	else
+		AM.layer = MOB_LAYER
+
+/datum/component/riding/human/get_offsets(pass_index)
+	var/mob/living/carbon/human/H = parent
+	if(H.buckle_lying)
+		return list(TEXT_NORTH = list(0, 6), TEXT_SOUTH = list(0, 6), TEXT_EAST = list(0, 6), TEXT_WEST = list(0, 6))
```
Comment (spookydonut) [inline code/datums/components/riding.dm]: what are these magic numbers, offsets?

File: code/modules/mob/living/carbon/human/human.dm [hunk 23]
```diff
@@ -787,6 +787,68 @@
...
+/mob/living/carbon/human/buckle_mob(mob/living/buckling_mob, force = FALSE, check_loc = TRUE, lying_buckle = FALSE, hands_needed = 0, target_hands_needed = 0, silent)
+	if(!force)//humans are only meant to be ridden through piggybacking and special cases
+		return
+	if(!is_type_in_typecache(buckling_mob, can_ride_typecache))
+		buckling_mob.visible_message("<span class='warning'>[buckling_mob] really can't seem to mount [src]...</span>")
+		return
+	buckle_lying = lying_buckle
+	var/datum/component/riding/human/riding_datum = LoadComponent(/datum/component/riding/human)
+	if(target_hands_needed)
+		riding_datum.ride_check_rider_restrained = TRUE
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/human/human.dm]: noooooooooooooooooooooooooooooooooooooooooooooooooo you shouldnt be changing component stuff like this

File: code/modules/mob/living/carbon/human/human.dm [hunk 24]
```diff
@@ -787,6 +787,68 @@
...
+	if(buckled || (buckling_mob in buckled_mobs) || (LAZYLEN(buckled_mobs) >= max_buckled_mobs))
+		return
+	var/equipped_hands_self
+	var/equipped_hands_target
+	if(hands_needed)
+		equipped_hands_self = riding_datum.equip_buckle_inhands(src, hands_needed, buckling_mob)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/human/human.dm]: same here yuck

File: code/modules/mob/living/carbon/human/human.dm [hunk 25]
```diff
@@ -787,6 +787,68 @@
...
+		visible_message("<span class='warning'>[src] can't get a grip on [buckling_mob] because their hands are full!</span>",
+			"<span class='warning'>You can't get a grip on [buckling_mob] because your hands are full!</span>")
+		return
+	else if(target_hands_needed && !equipped_hands_target)
+		buckling_mob.visible_message("<span class='warning'>[buckling_mob] can't get a grip on [src] because their hands are full!</span>",
+			"<span class='warning'>You can't get a grip on [src] because your hands are full!</span>")
+		return
+
+	stop_pulling()
+	riding_datum.handle_vehicle_layer()
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/human/human.dm]: yuck yuck yuck

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm (hunk 11 above)
Comment (spookydonut) [reply code/game/objects/structures/stool_bed_chair_nest/bed.dm]: BUCKLE_MOB_HORIZONTAL ?

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm (hunk 12 above)
Comment (spookydonut) [reply code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: BUCKLE_MOB_UPRIGHT ?

File: code/datums/components/riding.dm (hunk 22 above)
Comment (Rohesie) [reply code/datums/components/riding.dm]: Yea, the proc is called `get_offsets()`
Not my code, but not sure how that'd be more clear.

File: code/game/objects/effects/landmarks/corpsespawner.dm [hunk 26]
```diff
@@ -104,7 +104,8 @@
 		M.update_burst()
 		//buckle to nest
 		var/obj/structure/bed/nest/victim_nest = locate() in get_turf(src)
-		victim_nest?.do_buckle(M, silent = TRUE)
+		if(victim_nest)
+			victim_nest.buckle_mob(M, silent = TRUE)
```
Comment (psykzz) [inline code/game/objects/effects/landmarks/corpsespawner.dm]: Why this change, your normally against this type of thing?

File: code/game/objects/items/bodybag.dm [hunk 27]
```diff
@@ -175,7 +181,8 @@
 
 
 /obj/structure/closet/bodybag/forceMove(atom/destination)
-	roller_buckled?.unbuckle()
+	if(roller_buckled)
+		roller_buckled.unbuckle_bodybag()
```
Comment (psykzz) [inline code/game/objects/items/bodybag.dm]: Same here why the if instead of `roller_buckled?.unbuckle()	 `

File: code/game/objects/machinery/OpTable.dm [hunk 28]
```diff
@@ -12,8 +12,8 @@
 	active_power_usage = 5
 	var/mob/living/carbon/human/victim = null
 	var/strapped = 0.0
-	can_buckle = TRUE
-	buckle_lying = TRUE
+	buckle_flags = CAN_BUCKLE
+	buckle_lying = 90
```
Comment (psykzz) [inline code/game/objects/machinery/OpTable.dm]: Why 90?

Comment (psykzz) [reply code/game/objects/machinery/OpTable.dm]: I see this is now the angle.

File: code/datums/components/riding.dm (hunk 9 above)
Comment (Rohesie) [reply code/datums/components/riding.dm]: Done.

File: code/game/objects/effects/landmarks/corpsespawner.dm (hunk 26 above)
Comment (Rohesie) [reply code/game/objects/effects/landmarks/corpsespawner.dm]: Sounds nuts, but the profiling results show it, `var?.var` is more efficient than `var && var.var`, but `var?.proc()` is actually less than `var && var.proc()`

**Comment by GravitronBlue** [comment]
Heavy marine buff, stretchers are meant for recovery, not being able to just pick up and yakkety sax away with dead marines.
